react-instantsearch-hooks/geo-search/src/Airports.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { useEffect, useRef } from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { Marker, Popup, useMapEvents } from 'react-leaflet';
+import { DivIcon } from 'leaflet';
+import { useSearchBox } from 'react-instantsearch-hooks-web';
+import { useGeoSearch } from './useGeoSearch';
+
+import type { GeoHit } from 'instantsearch.js/es/connectors/geo-search/connectGeoSearch';
+
+export function Airports() {
+  const skipViewEffect = useRef(false);
+
+  const { query, refine: refineQuery } = useSearchBox();
+  const {
+    items,
+    refine: refineItems,
+    currentRefinement,
+    clearMapRefinement,
+  } = useGeoSearch();
+
+  const onViewChange = ({ target }) => {
+    skipViewEffect.current = true;
+
+    if (query.length > 0) {
+      refineQuery('');
+    }
+
+    refineItems({
+      northEast: target.getBounds().getNorthEast(),
+      southWest: target.getBounds().getSouthWest(),
+    });
+  };
+
+  const map = useMapEvents({
+    zoomend: onViewChange,
+    dragend: onViewChange,
+  });
+
+  useEffect(() => {
+    if (currentRefinement) {
+      clearMapRefinement();
+    }
+
+    if (items.length > 0 && !skipViewEffect.current) {
+      map.setView(items[0]._geoloc);
+    }
+
+    skipViewEffect.current = false;
+  }, [query]);
```

Comment:
The original intent is:

- to center the map on the first item when the query changes
- to refine on the viewable part of the map when it is moved / zoomed

The first part requires me to reset the boundaries refinement if necessary (`clearMapRefinement()`) when the query changes to allow searching for airports that are not in the currently viewable portion of the map.

The last part requires me to reset the query otherwise it will refine on results that match the query on top of the geographic coordinates. And so I needed to mark with a ref (`skipViewEffect`) this particular action (resetting the query due to the map being moved) in order not to recenter the map on the first item of a results list.

Reply:
I think it's fine as long as we add comments to the code here, since it's complex and it's supposed to be understandable by anyone.